src/distribution_spatial.cpp

```diff
@@ -180,6 +185,78 @@ Position SphericalIndependent::sample(uint64_t* seed) const
   return {x, y, z};
 }
 
+//==============================================================================
+// MeshIndependent implementation
+//==============================================================================
+
+MeshIndependent::MeshIndependent(pugi::xml_node node)
+{
+  // No in-tet distributions implemented, could include distributions for the barycentric coords
+  // Read in unstructured mesh from mesh_id value
+  int32_t mesh_id = std::stoi(get_node_value(node, "mesh_id"));
+  // Get pointer to spatial distribution
+  mesh_map_idx_ = model::mesh_map.at(mesh_id);
+  const auto& mesh_ptr = model::meshes[mesh_map_idx_];
+  
+  // Check whether mesh pointer points to an unstructured mesh
+  umesh_ptr_ = dynamic_cast<UnstructuredMesh*>(mesh_ptr.get());
+  if (!umesh_ptr_) {fatal_error("Mesh passed to spatial distribution is not an unstructured mesh object"); }
+
+  // Create CDF based on weighting scheme
+  tot_bins_ = umesh_ptr_->n_bins();
+  std::vector<double> weights = {};
+  weights.resize(tot_bins_);
+  double temp_total_weight = 0.0;
+  mesh_CDF_.resize(tot_bins_+1);
+  mesh_CDF_[0] = {0.0};
+  total_weight_ = 0.0;
+  mesh_weights_.resize(tot_bins_);
+
+  // Create cdfs for sampling for an element over a mesh
+  // Volume scheme is weighted by the volume of each tet
+  // File scheme is weighted by an array given in the xml file
+
+  if (check_for_node(node, "weights_from_file")) {
+    mesh_weights_ = get_node_array<double>(node, "weights_from_file");
+    if (mesh_weights_.size() != tot_bins_){
+      write_message("WARNING: The size of the weights array from the xml file does not equal the number of elements in the mesh.");
```

Comment:
I wondered whether this should be a warning or an error, in either case this line should use a real warning rather than just printing it out

Reply:
I used this mostly for testing, it was easier to use a small array and have everything else be zero rather than create a 12000 size array. Current tests should not be affected